### PR TITLE
[PATCH API-NEXT v4] crypto: add new algorithms

### PR DESCRIPTION
**Pull request by lumag**
This is a copy of #430 , incorporating changes from #428 , implementation of CCM (declared at #430) and declaration and implentation of AES-CMAC, requested by our customers. I'd like to get #422  in for TigerMoth. Otherwise we seem more or less complete.

### CODE REVIEW

File: test/validation/api/crypto/test_vectors.h [hunk 1]
```diff
@@ -444,6 +444,135 @@ static crypto_test_reference_t aes_gcm_reference[] = {
 	}
 };
 
+static crypto_test_reference_t aes_ccm_reference[] = {
+	/*
+	 * AES-CCM reference from RFC 3610:
+	 * https://tools.ietf.org/html/rfc3610#section-8
```
Comment (Bill-Fischofer-Linaro) [inline test/validation/api/crypto/test_vectors.h]: The RFC citation is sufficient. URLs should not be included as they tend to get stale.

File: test/validation/api/crypto/test_vectors.h [hunk 2]
```diff
@@ -477,6 +477,146 @@ static crypto_test_reference_t aes_gmac_reference[] = {
 	},
 };
 
+/*
+ * Test vector from RFC 7539, sections 2.8.2, A.5:
+ * https://tools.ietf.org/html/rfc7539#section-2.8.2
+ * https://tools.ietf.org/html/rfc7539#appendix-A.5
```
Comment (Bill-Fischofer-Linaro) [inline test/validation/api/crypto/test_vectors.h]: The RFC citation is sufficient.

File: test/validation/api/crypto/test_vectors.h [hunk 3]
```diff
@@ -477,6 +606,244 @@ static crypto_test_reference_t aes_gmac_reference[] = {
 	},
 };
 
+/* Test vectors from SP800-38B / CSRC examples:
+ * https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Standards-and-Guidelines/documents/examples/AES_CMAC.pdf
```
Comment (Bill-Fischofer-Linaro) [inline test/validation/api/crypto/test_vectors.h]: The NIST SP citation is sufficient. Any reader can easily Google it if needed. URLs tend to get stale quickly, and checkpatch flags it as too long.

**Comment by lumag** [comment]
#422 is not a prereq.
I will update PR, dropping URLs

**Comment by Bill-Fischofer-Linaro** [comment]
Remaining checkpatch issues are false positives.

**Comment by lumag** [comment]
@muvarov I'll refresh this PR after #436 is merged

**Comment by muvarov** [comment]
#436 is merged. You can refresh this one.

**Comment by lumag** [comment]
@muvarov @psavol rebased. Please review.

**Comment by muvarov** [comment]
Merged.

**Pull request closed without merging**